src/tmpfiles/tmpfiles.c

```diff
@@ -2006,62 +2006,176 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
+        }
+
+        if ((st.st_mode & S_IFMT) == S_IFDIR) {
+                r = rm_rf_children(fd, REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL, &st);
+                if (r < 0) {
+                        return r;
+                }
+
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(fd, name, AT_REMOVEDIR);
+        } else {
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(fd, name, 0);
+        }
         if (r < 0) {
                 if (errno != ENOENT) {
-                        log_error_errno(r, "rm_rf(%s): %m", path);
+                        // TODO fix this log to give full path like log_unsafe_transition
+                        log_error_errno(r, "rm_rf(%s): %m", name);
                 }
                 return -errno;
         }
+        return -ENOENT;
+}
 
-        return 0;
+static void close_parent(int* parent_fd) {
+        if (*parent_fd != AT_FDCWD) {
+                safe_close(*parent_fd);
+        }
+        *parent_fd = -1;
 }
 
-static int rm_nondir_parents(const char *path) {
-        char *p, *e;
-        int r = -ENOENT;
+/* If child_mode is non-zero, rm_if_wrong_type_safe will be executed for the last path component. */
+static int mkdir_parents_rm_if_wrong_type(mode_t child_mode, const char *path) {
+        _cleanup_(close_parent) int parent_fd = -1;
+        _cleanup_close_ int next_fd = -1;
+        struct stat parent_st;
+        const char *s, *e;
+        int r;
+        size_t path_len;
 
         assert(path);
+        assert((child_mode & ~S_IFMT) == 0);
 
-        // Walk up the path components until one is found with the expected type or there are no more.
-        p = strdupa(path);
-        // If the path doesn't exist, check the next path component.
-        while (r == -ENOENT || r == -ENOTDIR) {
-                e = strrchr(p, '/');
-                if (!e)
-                        return 0;
+        path_len = strlen(path);
+
+        e = strrchr(path, '/');
+        if (!e)
+                return child_mode ? rm_if_wrong_type_safe(child_mode, AT_FDCWD, NULL, path, false) : 0;
+
+        /* Trailing path separators are only allowed if child_mode isn't set. */
+        if ((e - path + 1 == path_len) && child_mode)
+                return -EINVAL;
+
+        /* Get the parent_fd and stat. */
+        parent_fd = AT_FDCWD;
+        if (*path == '/') {
+                parent_fd = open("/", O_NOCTTY | O_CLOEXEC | O_DIRECTORY);
+                // TODO log on error.
+                if (parent_fd < 0)
+                        return -errno;
+                r = fstat(parent_fd, &parent_st);
+        } else {
+                r = stat("", &parent_st);
+        }
+        // TODO log on error.
+        if (r < 0)
+                return -errno;
+
+        /* Check every parent directory in the path, except the last component */
+        e = path;
+        for (;;) {
```

Comment:
This is patterned after the mk_dir functions because I didn't find a macro for walking the path components in the file utils.